python/grass/tools/session_tools.py

```diff
@@ -0,0 +1,308 @@
+        if "env" not in popen_options:
+            popen_options["env"] = self._modified_env_if_needed()
+        if self._capture_output:
+            if "stdout" not in popen_options:
+                popen_options["stdout"] = gs.PIPE
+        if self._capture_stderr:
+            if "stderr" not in popen_options:
+                popen_options["stderr"] = gs.PIPE
+        if input is not None:
+            popen_options["stdin"] = gs.PIPE
+        else:
+            popen_options["stdin"] = None
+        process = gs.Popen(
+            command,
+            **popen_options,
+        )
+        stdout, stderr = process.communicate(input=input)
+        returncode = process.poll()
+        # We don't have the keyword arguments to pass to the resulting object.
+        result = ToolResult(
+            name=command[0],
+            command=command,
+            kwargs=tool_kwargs,
+            returncode=returncode,
+            stdout=stdout,
+            stderr=stderr,
+        )
+        if returncode != 0:
+            # This is only for the error states.
+            # The handle_errors function handles also the run_command functions
+            # and may use some overall review to make the handling of the tool name
+            # and parameters more clear, but currently, the first item in args is a
+            # list if it is a whole command.
+            args = [command[0]] if tool_kwargs else [command]
+            return gs.handle_errors(
+                returncode,
+                result=result,
+                args=args,
+                kwargs=tool_kwargs or {},
+                stderr=stderr,
+                handler=self._errors,
+            )
+        return result
+
+    def __getattr__(self, name):
+        """Get a function representing a GRASS tool.
+
+        Attribute should be in the form 'r_example_name'. For example, 'r.slope.aspect'
+        is used trough attribute 'r_slope_aspect'.
+        """
+        if not self._name_resolver:
+            self._name_resolver = ToolFunctionResolver(
+                run_function=self.run,
+                env=self._original_env,
+            )
+        return self._name_resolver.get_function(name, exception_type=AttributeError)
+
+    def __dir__(self):
+        """List available tools and standard attributes."""
+        if not self._name_resolver:
+            self._name_resolver = ToolFunctionResolver(
+                run_function=self.run,
+                env=self._original_env,
+            )
+        # Collect instance and class attributes
+        static_attrs = set(dir(type(self))) | set(self.__dict__.keys())
+        return list(static_attrs) + self._name_resolver.names()
+
+    def __enter__(self):
+        """Enter the context manager context.
+
+        :returns: reference to the object (self)
+        """
+        return self
+
+    def __exit__(self, type, value, traceback):
```

Comment:
The signature in the python docs is usually:
```suggestion
    def __exit__(self, exc_type, exc_value, traceback):
```

https://docs.python.org/3/reference/datamodel.html#object.__exit__
In typeshed, it is most often like that, but not always, like in  https://github.com/python/typeshed/blob/7c7f8cd184ccd96980477e56c7496d1d87784d1d/stdlib/contextlib.pyi

Otherwise, Pylint is complaining with [W0622 / redefined-builtin](https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/redefined-builtin.html) because of `type` that is a built-in function